Display full help with 'help' function

rkeithhill opened the pull request:
Display full help on Windows systems where user has overriden pager with $env:PAGER.

Fix #4962

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

rkeithhill committed:
Display full help on non-Windows systems

Display full help on Windows systems where user has overriden pager with $env:PAGER.

Fix #4962

rkeithhill commented:
Any chance of something like this making it in before 6.0.0 final?  It would be nice to offer full help to Linux users.  And to Windows  users who elect to set $env:PAGER to a better pager app.  :-)

adityapatwardhan commented:
@rkeithhill  Sorry for the delay. I will review this today.

rkeithhill commented:
No problem.  You all had a big push to get beta.9 out.

adityapatwardhan commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -4253,6 +4253,12 @@ .FORWARDHELPCATEGORY Cmdlet
+    # On Windows, only display the full help topic if the user has specified a PAGER app.
+    if (($psCmdlet.ParameterSetName -eq 'AllUsersView') -and !$Full -and (!$IsWindows -or (Test-Path env:PAGER))) {
+        $PSBoundParameters['Full'] = $true
+    }
```
Can we add a test which check the view that gets selected. It should have `#FullView`
```PowerShell
(Get-Help Get-Command -Full).psobject.typenames
```

rkeithhill replied:
Yes, I can add that.

rkeithhill replied:
BTW, to be clear, this tweak affects only the `help` function on Linux/macOS/Windows as well as the `man` alias on Windows.  That function pipes through a pager so we don't get rich objects out - just text.  That is why I'm using `-cmatch` to look for certain words that indicate it is full help.

rkeithhill committed:
Add tests for full view change to help function

adityapatwardhan commented on test/powershell/engine/Help/HelpSystem.Tests.ps1:
```diff
@@ -300,5 +300,83 @@ Describe "Get-Help should find pattern help files" -Tags "CI" {
+        }
+    }
+
+    It "help should not return full view without -Full switch" -Skip:(!$IsWindows) {
```
Since you already have and after all the `Skip` should be done over there instead of repeating it in every `It`.

https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md#skipping-tests-in-bulk

adityapatwardhan commented on test/powershell/engine/Help/HelpSystem.Tests.ps1:
```diff
@@ -300,5 +300,83 @@ Describe "Get-Help should find pattern help files" -Tags "CI" {
+        }
+    }
+
+    It "help should return full view without -Full switch" -Skip:(!$IsWindows) {
```
Same comment as above.

rkeithhill committed:
Addess PR feedback on skip approach

adityapatwardhan reviewed (approved):
LGTM

rkeithhill committed:
help function displays full help on all platforms

As per PowerShell Committee discussion - https://github.com/PowerShell/PowerShell/issues/4962#issuecomment-341265491

rkeithhill commented:
I think this is ready to go.  The feedback from the @PowerShell/powershell-committee review was implemented in the last commit.
